src/rest-server/src/middlewares/token.js

```diff
@@ -15,45 +15,53 @@
 // DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 
-
-const jwt = require('jsonwebtoken');
-
-const config = require('@pai/config/token');
+const {userProperty} = require('@pai/config/token');
+const userModel = require('@pai/models/v2/user');
+const tokenModel = require('@pai/models/token');
 const createError = require('@pai/utils/error');
 
-const createMiddleware = (throwErrorIfUnauthorized) => {
-  return function(req, _, next) {
-    try {
-      let token = getToken(req);
-      let result = jwt.verify(token, config.secret);
-      req[config.userProperty] = result;
-    } catch (error) {
-      if (error.name && error.name === 'TokenExpiredError') {
-        let error = createError('Unauthorized', 'UnauthorizedUserError', 'Your token is expired.');
-        return next(error);
-      }
-      if (throwErrorIfUnauthorized) {
-        let error = createError('Unauthorized', 'UnauthorizedUserError', 'Guest is not allowed to do this operation.');
-        return next(error);
-      }
+const getToken = async (req) => {
+  const [scheme, credentials] = req.headers.authorization.split(' ');
+  if (/^Basic$/i.test(scheme)) {
+    // application token
+    const raw = Buffer.from(credentials, 'base64').toString();
+    const [username, password] = raw.split(':');
+    const token = await tokenModel.verify(password);
+    if (token.username === username && token.application) {
+      return token;
+    } else {
+      throw new Error('Invalid authorization header');
     }
```

Comment:
Not quiet understand for this line. In which situation, the HTTP authentication header will be "Basic credentials"

Reply:
Other platforms ([github](https://developer.github.com/v3/auth/#basic-authentication), [vso](https://docs.microsoft.com/en-us/azure/devops/organizations/accounts/use-personal-access-tokens-to-authenticate?view=azure-devops#q-how-can-i-use-a-pat-in-my-code)) uses basic authentication for their pats (`BASIC username:pat_string`).
I provide a similar authentication option here if user is more used to them.

Reply:
Thanks, got it

---
In reply to: [339970596](https://github.com/microsoft/pai/pull/3774#discussion_r339970596) [](ancestors = 339970596)